// @handsontable/angular/esm5/handsontable-angular.js
import { Component, ElementRef, EventEmitter, Injectable, Injector, Input, NgModule, NgZone, Output, ViewEncapsulation } from '@angular/core';
import * as Handsontable from 'handsontable';
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var instances = new Map();
var HotTableRegisterer = /** @class */ (function () {
    function HotTableRegisterer() {
    }
    /**
     * @param {?} id
     * @return {?}
     */
    HotTableRegisterer.prototype.getInstance = function (id) {
        return instances.get(id);
    };
    /**
     * @param {?} id
     * @param {?} instance
     * @return {?}
     */
    HotTableRegisterer.prototype.registerInstance = function (id, instance) {
        instances.set(id, instance);
    };
    /**
     * @param {?} id
     * @return {?}
     */
    HotTableRegisterer.prototype.removeInstance = function (id) {
        instances.delete(id);
    };
    return HotTableRegisterer;
}());
HotTableRegisterer.decorators = [
    { type: Injectable },
];
/** @nocollapse */
HotTableRegisterer.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var AVAILABLE_OPTIONS = ['allowEmpty', 'allowHtml', 'allowInsertColumn', 'allowInsertRow',
    'allowInvalid', 'allowRemoveColumn', 'allowRemoveRow', 'autoColumnSize', 'autoComplete', 'autoRowSize',
    'autoWrapCol', 'autoWrapRow', 'cell', 'cells', 'checkedTemplate', 'className', 'colHeaders', 'columnHeaderHeight',
    'columns', 'columnSorting', 'colWidths', 'commentedCellClassName', 'comments', 'contextMenu', 'copyable',
    'copyPaste', 'correctFormat', 'currentColClassName', 'currentHeaderClassName', 'currentRowClassName',
    'customBorders', 'data', 'dataSchema', 'dateFormat', 'debug', 'defaultDate', 'disableVisualSelection',
    'dragToScroll', 'editor', 'enterBeginsEditing', 'enterMoves', 'fillHandle', 'filter', 'filteringCaseSensitive',
    'fixedColumnsLeft', 'fixedRowsTop', 'fragmentSelection', 'height', 'invalidCellClassName', 'label',
    'language', 'manualColumnFreeze', 'manualColumnMove', 'manualColumnResize', 'manualRowMove', 'manualRowResize',
    'maxCols', 'maxRows', 'mergeCells', 'minCols', 'minRows', 'minSpareCols', 'minSpareRows', 'multiSelect',
    'noWordWrapClassName', 'numericFormat', 'observeChanges', 'observeDOMVisibility', 'outsideClickDeselects',
    'persistentState', 'placeholder', 'placeholderCellClassName', 'preventOverflow', 'readOnly',
    'readOnlyCellClassName', 'renderAllRows', 'renderer', 'rowHeaders', 'rowHeaderWidth', 'rowHeights',
    'search', 'selectOptions', 'skipColumnOnPaste', 'sortByRelevance', 'sortFunction', 'sortIndicator',
    'source', 'startCols', 'startRows', 'stretchH', 'strict', 'tableClassName', 'tabMoves', 'title',
    'trimDropdown', 'trimWhitespace', 'type', 'uncheckedTemplate', 'undo', 'validator',
    'viewportColumnRenderingOffset', 'viewportRowRenderingOffset', 'visibleRows', 'width', 'wordWrap'];
var AVAILABLE_HOOKS = ['afterBeginEditing', 'afterCellMetaReset', 'afterChange',
    'afterChangesObserved', 'afterColumnMove', 'afterColumnResize', 'afterColumnSort',
    'afterContextMenuDefaultOptions', 'afterContextMenuHide', 'afterContextMenuShow', 'afterCopy',
    'afterCopyLimit', 'afterCreateCol', 'afterCreateRow', 'afterCut', 'afterDeselect', 'afterDestroy',
    'afterDocumentKeyDown', 'afterGetCellMeta', 'afterGetColHeader', 'afterGetColumnHeaderRenderers',
    'afterGetRowHeader', 'afterGetRowHeaderRenderers', 'afterInit', 'afterLanguageChange', 'afterListen',
    'afterLoadData', 'afterModifyTransformEnd', 'afterModifyTransformStart', 'afterMomentumScroll',
    'afterOnCellCornerDblClick', 'afterOnCellCornerMouseDown', 'afterOnCellMouseDown',
    'afterOnCellMouseOut', 'afterOnCellMouseOver', 'afterPaste', 'afterPluginsInitialized', 'afterRedo',
    'afterRemoveCellMeta', 'afterRemoveCol', 'afterRemoveRow', 'afterRender', 'afterRenderer',
    'afterRowMove', 'afterRowResize', 'afterScrollHorizontally', 'afterScrollVertically', 'afterSelection',
    'afterSelectionByProp', 'afterSelectionEnd', 'afterSelectionEndByProp', 'afterSetCellMeta',
    'afterSetDataAtCell', 'afterSetDataAtRowProp', 'afterUndo', 'afterUnlisten', 'afterUpdateSettings',
    'afterValidate', 'afterViewportColumnCalculatorOverride', 'afterViewportRowCalculatorOverride',
    'beforeAutofill', 'beforeAutofillInsidePopulate', 'beforeCellAlignment', 'beforeChange', 'beforeChangeRender',
    'beforeColumnMove', 'beforeColumnResize', 'beforeColumnSort', 'beforeContextMenuSetItems', 'beforeCopy',
    'beforeCreateCol', 'beforeCreateRow', 'beforeCut', 'beforeDrawBorders', 'beforeGetCellMeta', 'beforeInit',
    'beforeInitWalkontable', 'beforeKeyDown', 'beforeLanguageChange', 'beforeOnCellMouseDown', 'beforeOnCellMouseOut',
    'beforeOnCellMouseOver', 'beforePaste', 'beforeRedo', 'beforeRemoveCellMeta', 'beforeRemoveCol',
    'beforeRemoveRow', 'beforeRender', 'beforeRenderer', 'beforeRowMove', 'beforeRowResize', 'beforeSetRangeEnd',
    'beforeSetRangeEnd', 'beforeStretchingColumnWidth', 'beforeTouchScroll', 'beforeUndo', 'beforeValidate',
    'beforeValueRender', 'construct', 'init', 'manualRowHeights', 'modifyAutofillRange', 'modifyCol',
    'modifyColHeader', 'modifyColumnHeaderHeight', 'modifyColWidth', 'modifyCopyableRange', 'modifyData',
    'modifyRow', 'modifyRowData', 'modifyRowHeader', 'modifyRowHeaderWidth', 'modifyRowHeight', 'modifyTransformEnd',
    'modifyTransformStart', 'persistentStateLoad', 'persistentStateReset', 'persistentStateSave',
    'skipLengthCache', 'unmodifyCol', 'unmodifyRow'];
var HotSettingsResolver = /** @class */ (function () {
    function HotSettingsResolver() {
        this.hooks = AVAILABLE_HOOKS;
        this.options = AVAILABLE_OPTIONS;
    }
    /**
     * @param {?} component
     * @return {?}
     */
    HotSettingsResolver.prototype.mergeSettings = function (component) {
        var _this = this;
        var /** @type {?} */ mergedSettings = {};
        if (component['settings'] !== void 0) {
            Object.keys(component['settings']).forEach(function (key) {
                if (_this.hooks.indexOf(key) > -1) {
                    mergedSettings[key] = function (p1, p2, p3, p4, p5, p6) {
                        return component._ngZone.run(function () {
                            return component['settings'][key](p1, p2, p3, p4, p5, p6);
                        });
                    };
                }
                else {
                    mergedSettings[key] = component['settings'][key];
                }
            });
        }
        this.options.forEach(function (key) {
            var /** @type {?} */ option = component[key];
            if (option !== void 0) {
                mergedSettings[key] = option;
            }
        });
        this.hooks.forEach(function (key) {
            var /** @type {?} */ hook = component[key];
            if (hook && hook.observers.length > 0) {
                mergedSettings[key] = function (p1, p2, p3, p4, p5, p6) {
                    component._ngZone.run(function () {
                        component[key].emit({ hotInstance: component.hotInstance, params: [p1, p2, p3, p4, p5, p6] });
                    });
                };
            }
        });
        return mergedSettings;
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    HotSettingsResolver.prototype.prepareChanges = function (changes) {
        var /** @type {?} */ result = {};
        var /** @type {?} */ parameters = Object.keys(changes);
        parameters.forEach(function (param) {
            if (changes.hasOwnProperty(param)) {
                result[param] = changes[param].currentValue;
            }
        });
        return result;
    };
    return HotSettingsResolver;
}());
HotSettingsResolver.decorators = [
    { type: Injectable },
];
/** @nocollapse */
HotSettingsResolver.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var HotTableComponent = /** @class */ (function () {
    /**
     * @param {?} el
     * @param {?} _ngZone
     * @param {?} _hotTableRegisterer
     * @param {?} _hotSettingsResolver
     */
    function HotTableComponent(el, _ngZone, _hotTableRegisterer, _hotSettingsResolver) {
        this.el = el;
        this._ngZone = _ngZone;
        this._hotTableRegisterer = _hotTableRegisterer;
        this._hotSettingsResolver = _hotSettingsResolver;
        this.columnsComponents = [];
        this.afterBeginEditing = new EventEmitter();
        this.afterCellMetaReset = new EventEmitter();
        this.afterChange = new EventEmitter();
        this.afterChangesObserved = new EventEmitter();
        this.afterColumnMove = new EventEmitter();
        this.afterColumnResize = new EventEmitter();
        this.afterColumnSort = new EventEmitter();
        this.afterContextMenuDefaultOptions = new EventEmitter();
        this.afterContextMenuHide = new EventEmitter();
        this.afterContextMenuShow = new EventEmitter();
        this.afterCopy = new EventEmitter();
        this.afterCopyLimit = new EventEmitter();
        this.afterCreateCol = new EventEmitter();
        this.afterCreateRow = new EventEmitter();
        this.afterCut = new EventEmitter();
        this.afterDeselect = new EventEmitter();
        this.afterDestroy = new EventEmitter();
        this.afterDocumentKeyDown = new EventEmitter();
        this.afterGetCellMeta = new EventEmitter();
        this.afterGetColHeader = new EventEmitter();
        this.afterGetColumnHeaderRenderers = new EventEmitter();
        this.afterGetRowHeader = new EventEmitter();
        this.afterGetRowHeaderRenderers = new EventEmitter();
        this.afterInit = new EventEmitter();
        this.afterLanguageChange = new EventEmitter();
        this.afterListen = new EventEmitter();
        this.afterLoadData = new EventEmitter();
        this.afterModifyTransformEnd = new EventEmitter();
        this.afterModifyTransformStart = new EventEmitter();
        this.afterMomentumScroll = new EventEmitter();
        this.afterOnCellCornerDblClick = new EventEmitter();
        this.afterOnCellCornerMouseDown = new EventEmitter();
        this.afterOnCellMouseDown = new EventEmitter();
        this.afterOnCellMouseOut = new EventEmitter();
        this.afterOnCellMouseOver = new EventEmitter();
        this.afterPaste = new EventEmitter();
        this.afterPluginsInitialized = new EventEmitter();
        this.afterRedo = new EventEmitter();
        this.afterRemoveCellMeta = new EventEmitter();
        this.afterRemoveCol = new EventEmitter();
        this.afterRemoveRow = new EventEmitter();
        this.afterRender = new EventEmitter();
        this.afterRenderer = new EventEmitter();
        this.afterRowMove = new EventEmitter();
        this.afterRowResize = new EventEmitter();
        this.afterScrollHorizontally = new EventEmitter();
        this.afterScrollVertically = new EventEmitter();
        this.afterSelection = new EventEmitter();
        this.afterSelectionByProp = new EventEmitter();
        this.afterSelectionEnd = new EventEmitter();
        this.afterSelectionEndByProp = new EventEmitter();
        this.afterSetCellMeta = new EventEmitter();
        this.afterSetDataAtCell = new EventEmitter();
        this.afterSetDataAtRowProp = new EventEmitter();
        this.afterUndo = new EventEmitter();
        this.afterUnlisten = new EventEmitter();
        this.afterUpdateSettings = new EventEmitter();
        this.afterValidate = new EventEmitter();
        this.afterViewportColumnCalculatorOverride = new EventEmitter();
        this.afterViewportRowCalculatorOverride = new EventEmitter();
        this.beforeAutofill = new EventEmitter();
        this.beforeAutofillInsidePopulate = new EventEmitter();
        this.beforeCellAlignment = new EventEmitter();
        this.beforeChange = new EventEmitter();
        this.beforeChangeRender = new EventEmitter();
        this.beforeColumnMove = new EventEmitter();
        this.beforeColumnResize = new EventEmitter();
        this.beforeColumnSort = new EventEmitter();
        this.beforeContextMenuSetItems = new EventEmitter();
        this.beforeCopy = new EventEmitter();
        this.beforeCreateCol = new EventEmitter();
        this.beforeCreateRow = new EventEmitter();
        this.beforeCut = new EventEmitter();
        this.beforeDrawBorders = new EventEmitter();
        this.beforeGetCellMeta = new EventEmitter();
        this.beforeInit = new EventEmitter();
        this.beforeInitWalkontable = new EventEmitter();
        this.beforeKeyDown = new EventEmitter();
        this.beforeLanguageChange = new EventEmitter();
        this.beforeOnCellMouseDown = new EventEmitter();
        this.beforeOnCellMouseOut = new EventEmitter();
        this.beforeOnCellMouseOver = new EventEmitter();
        this.beforePaste = new EventEmitter();
        this.beforeRedo = new EventEmitter();
        this.beforeRemoveCellMeta = new EventEmitter();
        this.beforeRemoveCol = new EventEmitter();
        this.beforeRemoveRow = new EventEmitter();
        this.beforeRender = new EventEmitter();
        this.beforeRenderer = new EventEmitter();
        this.beforeRowMove = new EventEmitter();
        this.beforeRowResize = new EventEmitter();
        this.beforeSetRangeEnd = new EventEmitter();
        this.beforeSetRangeStart = new EventEmitter();
        this.beforeStretchingColumnWidth = new EventEmitter();
        this.beforeTouchScroll = new EventEmitter();
        this.beforeUndo = new EventEmitter();
        this.beforeValidate = new EventEmitter();
        this.beforeValueRender = new EventEmitter();
        this.construct = new EventEmitter();
        this.init = new EventEmitter();
        this.manualRowHeights = new EventEmitter();
        this.modifyAutofillRange = new EventEmitter();
        this.modifyCol = new EventEmitter();
        this.modifyColHeader = new EventEmitter();
        this.modifyColumnHeaderHeight = new EventEmitter();
        this.modifyColWidth = new EventEmitter();
        this.modifyCopyableRange = new EventEmitter();
        this.modifyData = new EventEmitter();
        this.modifyRow = new EventEmitter();
        this.modifyRowData = new EventEmitter();
        this.modifyRowHeader = new EventEmitter();
        this.modifyRowHeaderWidth = new EventEmitter();
        this.modifyRowHeight = new EventEmitter();
        this.modifyTransformEnd = new EventEmitter();
        this.modifyTransformStart = new EventEmitter();
        this.persistentStateLoad = new EventEmitter();
        this.persistentStateReset = new EventEmitter();
        this.persistentStateSave = new EventEmitter();
        this.skipLengthCache = new EventEmitter();
        this.unmodifyCol = new EventEmitter();
        this.unmodifyRow = new EventEmitter();
    }
    /**
     * @return {?}
     */
    HotTableComponent.prototype.ngOnInit = function () {
        this.container = document.createElement('div');
        if (this.hotId) {
            this.container.id = this.hotId;
        }
        this.el.nativeElement.appendChild(this.container);
    };
    /**
     * @return {?}
     */
    HotTableComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        var /** @type {?} */ options = this._hotSettingsResolver.mergeSettings(this);
        if (this.columnsComponents.length > 0) {
            var /** @type {?} */ columns_1 = [];
            this.columnsComponents.forEach(function (column) {
                columns_1.push(_this._hotSettingsResolver.mergeSettings(column));
            });
            options['columns'] = columns_1;
        }
        this._ngZone.runOutsideAngular(function () {
            _this.hotInstance = new Handsontable(_this.container, options);
        });
        if (this.hotId) {
            this._hotTableRegisterer.registerInstance(this.hotId, this.hotInstance);
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    HotTableComponent.prototype.ngOnChanges = function (changes) {
        if (this.hotInstance === void 0) {
            return;
        }
        var /** @type {?} */ newOptions = this._hotSettingsResolver.prepareChanges(changes);
        this.updateHotTable(newOptions);
    };
    /**
     * @return {?}
     */
    HotTableComponent.prototype.ngOnDestroy = function () {
        this.hotInstance.destroy();
        if (this.hotId) {
            this._hotTableRegisterer.removeInstance(this.hotId);
        }
        this.el.nativeElement.removeChild(this.container);
        this.container = void 0;
    };
    /**
     * @param {?} newSettings
     * @return {?}
     */
    HotTableComponent.prototype.updateHotTable = function (newSettings) {
        if (!this.hotInstance) {
            return;
        }
        this.hotInstance.updateSettings(newSettings, false);
    };
    /**
     * @return {?}
     */
    HotTableComponent.prototype.onAfterColumnsChange = function () {
        var _this = this;
        if (this.columnsComponents === void 0) {
            return;
        }
        if (this.columnsComponents.length > 0) {
            var /** @type {?} */ columns_2 = [];
            this.columnsComponents.forEach(function (column) {
                columns_2.push(_this._hotSettingsResolver.mergeSettings(column));
            });
            var /** @type {?} */ newOptions = {
                columns: columns_2
            };
            this.updateHotTable(newOptions);
        }
    };
    /**
     * @return {?}
     */
    HotTableComponent.prototype.onAfterColumnsNumberChange = function () {
        var _this = this;
        var /** @type {?} */ columns = [];
        if (this.columnsComponents.length > 0) {
            this.columnsComponents.forEach(function (column) {
                columns.push(_this._hotSettingsResolver.mergeSettings(column));
            });
        }
        this.updateHotTable({ columns: columns });
    };
    /**
     * @param {?} column
     * @return {?}
     */
    HotTableComponent.prototype.addColumn = function (column) {
        this.columnsComponents.push(column);
        this.onAfterColumnsNumberChange();
    };
    /**
     * @param {?} column
     * @return {?}
     */
    HotTableComponent.prototype.removeColumn = function (column) {
        var /** @type {?} */ index = this.columnsComponents.indexOf(column);
        this.columnsComponents.splice(index, 1);
        this.onAfterColumnsNumberChange();
    };
    return HotTableComponent;
}());
HotTableComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line
                selector: 'hot-table',
                template: '',
                encapsulation: ViewEncapsulation.None,
                providers: [HotTableRegisterer, HotSettingsResolver],
            },] },
];
/** @nocollapse */
HotTableComponent.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: NgZone, },
    { type: HotTableRegisterer, },
    { type: HotSettingsResolver, },
]; };
HotTableComponent.propDecorators = {
    "settings": [{ type: Input },],
    "hotId": [{ type: Input },],
    "allowEmpty": [{ type: Input },],
    "allowHtml": [{ type: Input },],
    "allowInsertColumn": [{ type: Input },],
    "allowInsertRow": [{ type: Input },],
    "allowInvalid": [{ type: Input },],
    "allowRemoveColumn": [{ type: Input },],
    "allowRemoveRow": [{ type: Input },],
    "autoColumnSize": [{ type: Input },],
    "autoComplete": [{ type: Input },],
    "autoRowSize": [{ type: Input },],
    "autoWrapCol": [{ type: Input },],
    "autoWrapRow": [{ type: Input },],
    "cell": [{ type: Input },],
    "cells": [{ type: Input },],
    "checkedTemplate": [{ type: Input },],
    "className": [{ type: Input },],
    "colHeaders": [{ type: Input },],
    "columnHeaderHeight": [{ type: Input },],
    "columns": [{ type: Input },],
    "columnSorting": [{ type: Input },],
    "colWidths": [{ type: Input },],
    "commentedCellClassName": [{ type: Input },],
    "comments": [{ type: Input },],
    "contextMenu": [{ type: Input },],
    "copyable": [{ type: Input },],
    "copyPaste": [{ type: Input },],
    "correctFormat": [{ type: Input },],
    "currentColClassName": [{ type: Input },],
    "currentHeaderClassName": [{ type: Input },],
    "currentRowClassName": [{ type: Input },],
    "customBorders": [{ type: Input },],
    "data": [{ type: Input },],
    "dataSchema": [{ type: Input },],
    "dateFormat": [{ type: Input },],
    "debug": [{ type: Input },],
    "defaultDate": [{ type: Input },],
    "disableVisualSelection": [{ type: Input },],
    "dragToScroll": [{ type: Input },],
    "editor": [{ type: Input },],
    "enterBeginsEditing": [{ type: Input },],
    "enterMoves": [{ type: Input },],
    "fillHandle": [{ type: Input },],
    "filter": [{ type: Input },],
    "filteringCaseSensitive": [{ type: Input },],
    "fixedColumnsLeft": [{ type: Input },],
    "fixedRowsTop": [{ type: Input },],
    "format": [{ type: Input },],
    "fragmentSelection": [{ type: Input },],
    "height": [{ type: Input },],
    "invalidCellClassName": [{ type: Input },],
    "label": [{ type: Input },],
    "language": [{ type: Input },],
    "manualColumnFreeze": [{ type: Input },],
    "manualColumnMove": [{ type: Input },],
    "manualColumnResize": [{ type: Input },],
    "manualRowMove": [{ type: Input },],
    "manualRowResize": [{ type: Input },],
    "maxCols": [{ type: Input },],
    "maxRows": [{ type: Input },],
    "mergeCells": [{ type: Input },],
    "minCols": [{ type: Input },],
    "minRows": [{ type: Input },],
    "minSpareCols": [{ type: Input },],
    "minSpareRows": [{ type: Input },],
    "multiSelect": [{ type: Input },],
    "noWordWrapClassName": [{ type: Input },],
    "numericFormat": [{ type: Input },],
    "observeChanges": [{ type: Input },],
    "observeDOMVisibility": [{ type: Input },],
    "outsideClickDeselects": [{ type: Input },],
    "pasteMode": [{ type: Input },],
    "persistentState": [{ type: Input },],
    "placeholder": [{ type: Input },],
    "placeholderCellClassName": [{ type: Input },],
    "preventOverflow": [{ type: Input },],
    "readOnly": [{ type: Input },],
    "readOnlyCellClassName": [{ type: Input },],
    "renderAllRows": [{ type: Input },],
    "renderer": [{ type: Input },],
    "rowHeaders": [{ type: Input },],
    "rowHeaderWidth": [{ type: Input },],
    "rowHeights": [{ type: Input },],
    "search": [{ type: Input },],
    "selectOptions": [{ type: Input },],
    "skipColumnOnPaste": [{ type: Input },],
    "sortByRelevance": [{ type: Input },],
    "sortFunction": [{ type: Input },],
    "sortIndicator": [{ type: Input },],
    "source": [{ type: Input },],
    "startCols": [{ type: Input },],
    "startRows": [{ type: Input },],
    "stretchH": [{ type: Input },],
    "strict": [{ type: Input },],
    "tableClassName": [{ type: Input },],
    "tabMoves": [{ type: Input },],
    "title": [{ type: Input },],
    "trimDropdown": [{ type: Input },],
    "trimWhitespace": [{ type: Input },],
    "type": [{ type: Input },],
    "uncheckedTemplate": [{ type: Input },],
    "undo": [{ type: Input },],
    "validator": [{ type: Input },],
    "viewportColumnRenderingOffset": [{ type: Input },],
    "viewportRowRenderingOffset": [{ type: Input },],
    "visibleRows": [{ type: Input },],
    "width": [{ type: Input },],
    "wordWrap": [{ type: Input },],
    "afterBeginEditing": [{ type: Output },],
    "afterCellMetaReset": [{ type: Output },],
    "afterChange": [{ type: Output },],
    "afterChangesObserved": [{ type: Output },],
    "afterColumnMove": [{ type: Output },],
    "afterColumnResize": [{ type: Output },],
    "afterColumnSort": [{ type: Output },],
    "afterContextMenuDefaultOptions": [{ type: Output },],
    "afterContextMenuHide": [{ type: Output },],
    "afterContextMenuShow": [{ type: Output },],
    "afterCopy": [{ type: Output },],
    "afterCopyLimit": [{ type: Output },],
    "afterCreateCol": [{ type: Output },],
    "afterCreateRow": [{ type: Output },],
    "afterCut": [{ type: Output },],
    "afterDeselect": [{ type: Output },],
    "afterDestroy": [{ type: Output },],
    "afterDocumentKeyDown": [{ type: Output },],
    "afterGetCellMeta": [{ type: Output },],
    "afterGetColHeader": [{ type: Output },],
    "afterGetColumnHeaderRenderers": [{ type: Output },],
    "afterGetRowHeader": [{ type: Output },],
    "afterGetRowHeaderRenderers": [{ type: Output },],
    "afterInit": [{ type: Output },],
    "afterLanguageChange": [{ type: Output },],
    "afterListen": [{ type: Output },],
    "afterLoadData": [{ type: Output },],
    "afterModifyTransformEnd": [{ type: Output },],
    "afterModifyTransformStart": [{ type: Output },],
    "afterMomentumScroll": [{ type: Output },],
    "afterOnCellCornerDblClick": [{ type: Output },],
    "afterOnCellCornerMouseDown": [{ type: Output },],
    "afterOnCellMouseDown": [{ type: Output },],
    "afterOnCellMouseOut": [{ type: Output },],
    "afterOnCellMouseOver": [{ type: Output },],
    "afterPaste": [{ type: Output },],
    "afterPluginsInitialized": [{ type: Output },],
    "afterRedo": [{ type: Output },],
    "afterRemoveCellMeta": [{ type: Output },],
    "afterRemoveCol": [{ type: Output },],
    "afterRemoveRow": [{ type: Output },],
    "afterRender": [{ type: Output },],
    "afterRenderer": [{ type: Output },],
    "afterRowMove": [{ type: Output },],
    "afterRowResize": [{ type: Output },],
    "afterScrollHorizontally": [{ type: Output },],
    "afterScrollVertically": [{ type: Output },],
    "afterSelection": [{ type: Output },],
    "afterSelectionByProp": [{ type: Output },],
    "afterSelectionEnd": [{ type: Output },],
    "afterSelectionEndByProp": [{ type: Output },],
    "afterSetCellMeta": [{ type: Output },],
    "afterSetDataAtCell": [{ type: Output },],
    "afterSetDataAtRowProp": [{ type: Output },],
    "afterUndo": [{ type: Output },],
    "afterUnlisten": [{ type: Output },],
    "afterUpdateSettings": [{ type: Output },],
    "afterValidate": [{ type: Output },],
    "afterViewportColumnCalculatorOverride": [{ type: Output },],
    "afterViewportRowCalculatorOverride": [{ type: Output },],
    "beforeAutofill": [{ type: Output },],
    "beforeAutofillInsidePopulate": [{ type: Output },],
    "beforeCellAlignment": [{ type: Output },],
    "beforeChange": [{ type: Output },],
    "beforeChangeRender": [{ type: Output },],
    "beforeColumnMove": [{ type: Output },],
    "beforeColumnResize": [{ type: Output },],
    "beforeColumnSort": [{ type: Output },],
    "beforeContextMenuSetItems": [{ type: Output },],
    "beforeCopy": [{ type: Output },],
    "beforeCreateCol": [{ type: Output },],
    "beforeCreateRow": [{ type: Output },],
    "beforeCut": [{ type: Output },],
    "beforeDrawBorders": [{ type: Output },],
    "beforeGetCellMeta": [{ type: Output },],
    "beforeInit": [{ type: Output },],
    "beforeInitWalkontable": [{ type: Output },],
    "beforeKeyDown": [{ type: Output },],
    "beforeLanguageChange": [{ type: Output },],
    "beforeOnCellMouseDown": [{ type: Output },],
    "beforeOnCellMouseOut": [{ type: Output },],
    "beforeOnCellMouseOver": [{ type: Output },],
    "beforePaste": [{ type: Output },],
    "beforeRedo": [{ type: Output },],
    "beforeRemoveCellMeta": [{ type: Output },],
    "beforeRemoveCol": [{ type: Output },],
    "beforeRemoveRow": [{ type: Output },],
    "beforeRender": [{ type: Output },],
    "beforeRenderer": [{ type: Output },],
    "beforeRowMove": [{ type: Output },],
    "beforeRowResize": [{ type: Output },],
    "beforeSetRangeEnd": [{ type: Output },],
    "beforeSetRangeStart": [{ type: Output },],
    "beforeStretchingColumnWidth": [{ type: Output },],
    "beforeTouchScroll": [{ type: Output },],
    "beforeUndo": [{ type: Output },],
    "beforeValidate": [{ type: Output },],
    "beforeValueRender": [{ type: Output },],
    "construct": [{ type: Output },],
    "init": [{ type: Output },],
    "manualRowHeights": [{ type: Output },],
    "modifyAutofillRange": [{ type: Output },],
    "modifyCol": [{ type: Output },],
    "modifyColHeader": [{ type: Output },],
    "modifyColumnHeaderHeight": [{ type: Output },],
    "modifyColWidth": [{ type: Output },],
    "modifyCopyableRange": [{ type: Output },],
    "modifyData": [{ type: Output },],
    "modifyRow": [{ type: Output },],
    "modifyRowData": [{ type: Output },],
    "modifyRowHeader": [{ type: Output },],
    "modifyRowHeaderWidth": [{ type: Output },],
    "modifyRowHeight": [{ type: Output },],
    "modifyTransformEnd": [{ type: Output },],
    "modifyTransformStart": [{ type: Output },],
    "persistentStateLoad": [{ type: Output },],
    "persistentStateReset": [{ type: Output },],
    "persistentStateSave": [{ type: Output },],
    "skipLengthCache": [{ type: Output },],
    "unmodifyCol": [{ type: Output },],
    "unmodifyRow": [{ type: Output },],
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var HotColumnComponent = /** @class */ (function () {
    /**
     * @param {?} inj
     */
    function HotColumnComponent(inj) {
        this.inj = inj;
        this.firstRun = true;
        this.parentComponent = this.inj.get(HotTableComponent);
    }
    /**
     * @return {?}
     */
    HotColumnComponent.prototype.ngOnInit = function () {
        this.firstRun = false;
        this.parentComponent.addColumn(this);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    HotColumnComponent.prototype.ngOnChanges = function (changes) {
        if (this.firstRun) {
            return;
        }
        this.parentComponent.onAfterColumnsChange();
    };
    /**
     * @return {?}
     */
    HotColumnComponent.prototype.ngOnDestroy = function () {
        this.parentComponent.removeColumn(this);
    };
    return HotColumnComponent;
}());
HotColumnComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line
                selector: 'hot-column',
                template: '',
            },] },
];
/** @nocollapse */
HotColumnComponent.ctorParameters = function () { return [
    { type: Injector, },
]; };
HotColumnComponent.propDecorators = {
    "allowEmpty": [{ type: Input },],
    "allowHtml": [{ type: Input },],
    "allowInsertColumn": [{ type: Input },],
    "allowInsertRow": [{ type: Input },],
    "allowInvalid": [{ type: Input },],
    "allowRemoveColumn": [{ type: Input },],
    "allowRemoveRow": [{ type: Input },],
    "autoColumnSize": [{ type: Input },],
    "autoComplete": [{ type: Input },],
    "autoRowSize": [{ type: Input },],
    "autoWrapCol": [{ type: Input },],
    "autoWrapRow": [{ type: Input },],
    "cell": [{ type: Input },],
    "cells": [{ type: Input },],
    "checkedTemplate": [{ type: Input },],
    "className": [{ type: Input },],
    "colHeaders": [{ type: Input },],
    "columnHeaderHeight": [{ type: Input },],
    "columns": [{ type: Input },],
    "columnSorting": [{ type: Input },],
    "colWidths": [{ type: Input },],
    "commentedCellClassName": [{ type: Input },],
    "comments": [{ type: Input },],
    "contextMenu": [{ type: Input },],
    "copyable": [{ type: Input },],
    "copyPaste": [{ type: Input },],
    "correctFormat": [{ type: Input },],
    "currentColClassName": [{ type: Input },],
    "currentHeaderClassName": [{ type: Input },],
    "currentRowClassName": [{ type: Input },],
    "customBorders": [{ type: Input },],
    "data": [{ type: Input },],
    "dataSchema": [{ type: Input },],
    "dateFormat": [{ type: Input },],
    "debug": [{ type: Input },],
    "defaultDate": [{ type: Input },],
    "disableVisualSelection": [{ type: Input },],
    "dragToScroll": [{ type: Input },],
    "editor": [{ type: Input },],
    "enterBeginsEditing": [{ type: Input },],
    "enterMoves": [{ type: Input },],
    "fillHandle": [{ type: Input },],
    "filter": [{ type: Input },],
    "filteringCaseSensitive": [{ type: Input },],
    "fixedColumnsLeft": [{ type: Input },],
    "fixedRowsTop": [{ type: Input },],
    "fragmentSelection": [{ type: Input },],
    "height": [{ type: Input },],
    "invalidCellClassName": [{ type: Input },],
    "label": [{ type: Input },],
    "language": [{ type: Input },],
    "manualColumnFreeze": [{ type: Input },],
    "manualColumnMove": [{ type: Input },],
    "manualColumnResize": [{ type: Input },],
    "manualRowMove": [{ type: Input },],
    "manualRowResize": [{ type: Input },],
    "maxCols": [{ type: Input },],
    "maxRows": [{ type: Input },],
    "mergeCells": [{ type: Input },],
    "minCols": [{ type: Input },],
    "minRows": [{ type: Input },],
    "minSpareCols": [{ type: Input },],
    "minSpareRows": [{ type: Input },],
    "multiSelect": [{ type: Input },],
    "noWordWrapClassName": [{ type: Input },],
    "numericFormat": [{ type: Input },],
    "observeChanges": [{ type: Input },],
    "observeDOMVisibility": [{ type: Input },],
    "outsideClickDeselects": [{ type: Input },],
    "persistentState": [{ type: Input },],
    "placeholder": [{ type: Input },],
    "placeholderCellClassName": [{ type: Input },],
    "preventOverflow": [{ type: Input },],
    "readOnly": [{ type: Input },],
    "readOnlyCellClassName": [{ type: Input },],
    "renderAllRows": [{ type: Input },],
    "renderer": [{ type: Input },],
    "rowHeaders": [{ type: Input },],
    "rowHeaderWidth": [{ type: Input },],
    "rowHeights": [{ type: Input },],
    "search": [{ type: Input },],
    "selectOptions": [{ type: Input },],
    "skipColumnOnPaste": [{ type: Input },],
    "sortByRelevance": [{ type: Input },],
    "sortFunction": [{ type: Input },],
    "sortIndicator": [{ type: Input },],
    "source": [{ type: Input },],
    "startCols": [{ type: Input },],
    "startRows": [{ type: Input },],
    "stretchH": [{ type: Input },],
    "strict": [{ type: Input },],
    "tableClassName": [{ type: Input },],
    "tabMoves": [{ type: Input },],
    "title": [{ type: Input },],
    "trimDropdown": [{ type: Input },],
    "trimWhitespace": [{ type: Input },],
    "type": [{ type: Input },],
    "uncheckedTemplate": [{ type: Input },],
    "undo": [{ type: Input },],
    "validator": [{ type: Input },],
    "viewportColumnRenderingOffset": [{ type: Input },],
    "viewportRowRenderingOffset": [{ type: Input },],
    "visibleRows": [{ type: Input },],
    "width": [{ type: Input },],
    "wordWrap": [{ type: Input },],
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var HotTableModule = /** @class */ (function () {
    function HotTableModule() {
    }
    /**
     * @return {?}
     */
    HotTableModule.forRoot = function () {
        return {
            ngModule: HotTableModule,
            providers: [
                HotTableRegisterer
            ]
        };
    };
    return HotTableModule;
}());
HotTableModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    HotTableComponent,
                    HotColumnComponent,
                ],
                exports: [
                    HotTableComponent,
                    HotColumnComponent,
                ]
            },] },
];
/** @nocollapse */
HotTableModule.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { HotTableComponent, HotColumnComponent, HotTableRegisterer, HotSettingsResolver, HotTableModule };
